Add some grep help in reports dispatcher

proteusvacuum opened the pull request:
## Technical Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->
It's always impossible to find where reports are generating their URLs. Short of rewriting this logic, I thought this was at least slightly better.
https://dimagi.slack.com/archives/CP7S2N58W/p1642702627001400

## Safety Assurance
Comments only

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.
-->

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->


<!-- Please link to any past code changes that are coordinated with this migration -->

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

proteusvacuum committed:
Add some grep help in reports dispatcher

mkangia commented on corehq/apps/reports/dispatcher.py:
```diff
@@ -184,6 +193,15 @@ def _rendering_pattern(cls):
+        # ^(email/)?(?P<report_slug>[\w_]+)/$
+        # ^(partial/)?(?P<report_slug>[\w_]+)/$
+        # ^(print/)?(?P<report_slug>[\w_]+)/$
         return r'^({renderings}/)?(?P<report_slug>[\w_]+)/$'.format(renderings=cls._rendering_pattern())
```
I think it would be better to write a short documentation where this would be relevant rather than adding routes in the code. If I am looking for a view for a url, I don't see myself searching for the url regex in the code and finding this anyway.

proteusvacuum replied:
How do you generally find view URLs?
I normally search for a keyword in the URL, along with the app name that I think it would be in.

## Before
![Screenshot from 2022-01-25 09-59-09](https://user-images.githubusercontent.com/146896/151001336-d408b587-96c8-42b3-901d-a18184bdc5f9.png)
## After
![Screenshot from 2022-01-25 09-59-43](https://user-images.githubusercontent.com/146896/151001351-7fea62d3-b62c-460e-ba7f-b5682b2016c8.png)

I'm not trying to solve this problem for every single workflow, just those that use grep.

mkangia replied:
> How do you generally find view URLs?
I normally search for a keyword in the URL, along with the app name that I think it would be in.

Quite similar. I look for the keyword in just the files ending with name `urls.py` to limit my search.

esoergel replied:
I usually use `./manage.py show_urls | grep mykeyword` (from django-extensions).  Whenever I've tried that with reports I just sigh and open the dispatcher code.  I wonder how hard it'd be to dynamically register these URLs with Django (could even do it for each report), which I think would "fix" that workflow (clearly outside the scope of this PR)

mkangia commented on corehq/apps/reports/dispatcher.py:
```diff
@@ -153,6 +153,15 @@ def dispatch(self, request, domain=None, report_slug=None, render_as=None,
+                # return report.mobile_response
+                # return report.email_response
+                # return report.partial_response
+                # return report.print_response
```
curious to what our recommended replacement should be for such `attr` calls?
Should this instead be replaced with something like
```
if render_as == 'json':
    return report.json_response
elif render_as == 'async':
    return report.async_response
    # ...
```
or you could instead do
```
report.get_response(render_as)
```
and have report.get_response do the attr call, which I believe will be a touch better or do the if-else chain.

sounds like something book-club would be interested to weigh in on.

mkangia replied:
again, this documentation would not really help. People who use IDE wont benefit from this, and searching for say "json_response" would lead to a lot of matches in the code.

proteusvacuum replied:
Can you elaborate on why people who use an IDE wouldn't benefit? 

If I'm searching where `export_response` is called in the code, I would do something like: 
![Screenshot from 2022-01-25 10-13-40](https://user-images.githubusercontent.com/146896/151003881-b3ba4a70-d76d-4b77-8d2e-7a742bc3515d.png)

Prior to this change, when you search for that, it looks like the only place that method is called is in `reports/views.py`. This is very confusing, because that isn't where most calls to that method come from, and the only way I was able to figure it out was to put in a breakpoint in the method and then search up the call stack. 

Does your IDE know that this method is called here? If so, that's awesome - however, I don't think adding this line will hinder that usecase?

mkangia replied:
in IDE, I'd search to find "usages" of a method which the IDE provides which isn't same as text search but actual code calls. The change here seems beneficial for if you were doing text search and that too for "export_response" but if you search for "json_response", I am guessing you would get quite some irrelevant matches.

> Does your IDE know that this method is called here? If so, that's awesome

It can't find indirect calls like this.

I am not saying that there isn't currently a problem. I am just saying that the fix here isn't able to really solve it.

proteusvacuum replied:
OK... But do you think this is at least slightly better than not having these comments?

mkangia replied:
Sure yea. 

My concern is that adding these comments will make us think the problem is solved and accept this as the solution which I think isn't a good path to go down to.

esoergel replied:
This behavior has long annoyed me.  To @mkangia 's first point, another approach would be:
```python
return {
    'json': report.json_response,
    'async': report.async_response,
    ...
}[render_as]
```
But those are `@property`s that do work, a lot of it (another antipattern), and this approach would evaluate them all.  You could hack around this like so:
```python
return {
    'json': lambda: report.json_response,
    'async': lambda: report.async_response,
    ...
}[render_as]()
```
This would work too, but it's only marginally more discoverable than the original version:
```python
response_method = {
    'json': `json_response`,
    'async': `async_response`,
    ...
}[render_as]
return getattr(report, response_method)
```

kaapstorm replied:
My vote would be for Option Two: the dictionary with lambdas.

* It addresses the same problem that the grep help is addressing (although it does miss "return")
* It works for finding usages in PyCharm and VS Code
* Although the lambdas aren't ideal, it feels more concise than a 16-line `if ... elif` block, but still readable.
* Devs will be familiar with the dict lookup from other places in the codebase.

kaapstorm reviewed (approved):
:+1: Definitely moves the needle in the right direction. Progress is iterative.

kaapstorm committed:
isort

kaapstorm committed:
Setting a string as a default value is fine.

kaapstorm committed:
Official backports for type hints

kaapstorm committed:
Add type hints for devs who use IDEs

Add type hints for devs who use IDEs and devs who find type hints to be
a useful form of documentation.

kaapstorm committed:
Replace getattr with dict lookup

kaapstorm committed:
Resolve "TODO"

kaapstorm committed:
Explain why

kaapstorm committed:
OCD

kaapstorm committed:
Don't try. Do.

kaapstorm committed:
Explain dependency

kaapstorm committed:
Merge pull request #31024 from dimagi/nh/ide_help

Add some IDE help in reports dispatcher

kaapstorm committed:
Merge branch 'master' into fr/boo-meta-programming

kaapstorm committed:
Merge branch 'master' into fr/boo-meta-programming

kaapstorm committed:
Fix tests

kaapstorm committed:
An assertion is more helpful than a key error

kaapstorm committed:
Drop typing_extensions

This reverts commit 08f2693e9e74560221b50df2df1a1f615565a54f.

All environments are on Python 3.9
